### PR TITLE
add a nonblocking version of PostLocal CORE-4039

### PR DESCRIPTION
**Pull request by mmaxim**
@maxtaco r?

This adds the nonblocking version of `PostLocal` to the service, and a new `Outbox` and `Deliverer` to make it happen. I'll probably just need to walk through this with you, so that is all I will say 😄

### CODE REVIEW

**Comment by mmaxim** [comment]
@chrisnojima cc

**Comment by chrisnojima** [comment]
@mmaxim fixed

File: go/chat/storage/outbox.go [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+}
+
+func (o *Outbox) dbKey() libkb.DbKey {
+	return libkb.DbKey{
+		Typ: libkb.DBChatOutbox,
+		Key: "ob",
+	}
+}
+
+func (o *Outbox) readBox() ([]chat1.OutboxRecord, error) {
```
Comment (mmaxim) [inline go/chat/storage/outbox.go]: Add comments to make purpose clear (or change name)

File: go/chat/storage/outbox.go [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+	}
+	var res []chat1.OutboxRecord
+	if err = decode(pt, &res); err != nil {
+		return []chat1.OutboxRecord{}, err
+	}
+
+	return res, nil
+}
+
+func (o *Outbox) writeBox(outbox []chat1.OutboxRecord) error {
```
Comment (mmaxim) [inline go/chat/storage/outbox.go]: Same deal here for name.

File: go/chat/storage/outbox.go [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+func (o *Outbox) maybeNuke(err libkb.ChatStorageError) libkb.ChatStorageError {
+	if err.ShouldClear() {
+		if err := o.G().LocalChatDb.Delete(o.dbKey()); err != nil {
+			o.G().Log.Error("unable to clear inbox on error! err: %s", err.Error())
+		}
+	}
+	return err
+}
+
+func (o *Outbox) Push(convID chat1.ConversationID, msg chat1.MessagePlaintext) (chat1.OutboxID, libkb.ChatStorageError) {
```
Comment (mmaxim) [inline go/chat/storage/outbox.go]: Call `PushMessage`

File: go/chat/storage/outbox.go [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+	// Write out box
+	if err := o.writeBox(obox); err != nil {
+		return nil, o.maybeNuke(libkb.NewChatStorageInternalError(o.G(),
+			"error writing outbox: err: %s", err.Error()))
+	}
+
+	return outboxID, nil
+}
+
+func (o *Outbox) Pull() ([]chat1.OutboxRecord, error) {
```
Comment (mmaxim) [inline go/chat/storage/outbox.go]: Call `PullAllConversations`

File: go/chat/storage/outbox.go [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+	obox, err := o.readBox()
+	if err != nil {
+		return nil, o.maybeNuke(libkb.NewChatStorageInternalError(o.G(),
+			"error reading outbox: err: %s", err.Error()))
+	}
+
+	return obox, nil
+}
+
+func (o *Outbox) PullConv(convID chat1.ConversationID) ([]chat1.OutboxRecord, error) {
```
Comment (mmaxim) [inline go/chat/storage/outbox.go]: Call `PullConversation`

File: go/chat/storage/outbox.go [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+	var res []chat1.OutboxRecord
+	for _, obr := range obox {
+		if obr.ConvID.Eq(convID) {
+			res = append(res, obr)
+		}
+	}
+	return res, nil
+}
+
+func (o *Outbox) PopN(n int) error {
```
Comment (mmaxim) [inline go/chat/storage/outbox.go]: Call `PopNOldestMessages`

File: go/chat/sender.go [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+	}
+
+	// For now, BoxMessage canonicalizes the TLF name. We should try to refactor
+	// it a bit to do it here.
+	boxed, err := s.boxer.BoxMessage(ctx, msg, skp)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: populate plaintext.ClientHeader.Conv
```
Comment (mmaxim) [inline go/chat/sender.go]: Look into this

File: go/chat/sender.go [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+	"github.com/keybase/client/go/protocol/keybase1"
+	"golang.org/x/net/context"
+)
+
+type Sender interface {
+	Send(ctx context.Context, convID chat1.ConversationID, msg chat1.MessagePlaintext) (chat1.OutboxID, *chat1.RateLimit, error)
+	Prepare(ctx context.Context, msg chat1.MessagePlaintext, convID *chat1.ConversationID) (*chat1.MessageBoxed, error)
+}
+
+type BaseSender struct {
```
Comment (mmaxim) [inline go/chat/sender.go]: Call `BlockingSender`

File: go/libkb/globals.go [hunk 9]
```diff
@@ -88,7 +88,9 @@ type GlobalContext struct {
 
 	CardCache *UserCardCache // cache of keybase1.UserCard objects
 
-	ConvSource ConversationSource // source of remote message bodies for chat
+	ConvSource            ConversationSource // source of remote message bodies for chat
+	MessageDeliverer      MessageDeliverer   // background message delivery service
+	StartMessageDeliverer sync.Once
```
Comment (mmaxim) [inline go/libkb/globals.go]: Try to get this out of here.

**Comment by mmaxim** [comment]
Need changes to Outbox for logged in UID:
1.) Outbox needs to be parameteized by UID
2.) Deliverer needs to deliver on behalf of user.
3.) Start Deliverer from service/main.go
4.) Handle login/logout like Gregor.

**Comment by mmaxim** [comment]
@maxtaco PTAL

File: go/chat/sender.go (hunk 7 above)
Comment (maxtaco) [inline go/chat/sender.go]: Any thoughts on this `TODO`?

File: go/chat/sender.go [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+	}
+
+	// TODO: populate plaintext.ClientHeader.Conv
+
+	return boxed, nil
+}
+
+func (s *BlockingSender) getSigningKeyPair() (kp libkb.NaclSigningKeyPair, err error) {
+	// get device signing key for this user
+	signingKey, err := engine.GetMySecretKey(s.G(), s.getSecretUI, libkb.DeviceSigningKeyType, "sign chat message")
```
Comment (maxtaco) [inline go/chat/sender.go]: What will happen if we actually need a passphrase from the background?  There's nothing possible to do here, is there?

Comment (mmaxim) [inline go/chat/sender.go]: Don't think there is anything we can do, but it is not a big problem, it just means for that iteration of the `deliverLoop` we don't send anything. Once the key is unlocked we can send again.

Comment (mmaxim) [inline go/chat/sender.go]: Or once the key is cached we can send again.

Comment (mmaxim) [inline go/chat/sender.go]: As far as the `TODO`, I just removed it, that field is set in `BoxMessage`

**Comment by maxtaco** [comment]
Ok, good to go!

**Pull request merged**